File: typescript/list-ops/list-ops.ts
// Do *not* construct any array literal ([]) in your solution.
// Do *not* construct any arrays through new Array in your solution.
// DO *not* use any of the Array.prototype methods in your solution.

// You may use the destructuring and spreading (...) syntax from Iterable.

export class List {

  constructor(readonly values: unknown[]) { }

  public static create(...values: unknown[]): List {
    return new List(values)
  }

  public forEach(callback: (e: unknown) => void): void {
    for (const item of this.values) {
      callback(item);
    }
  }

  public append(list: List): List {
    return List.create(...this.values, ...list.values)
  }

  public concat(listOfLists: List): List {
    return listOfLists.foldl((acc: List, curr: List) => acc.append(curr), List.create(...this.values))
  }
  
  public filter<T>(predicate: (e: T) => boolean): List {
    return this.foldr((acc: List, curr: T) => predicate(curr) ? List.create(curr, ...acc.values) : acc, List.create()) 
  }

  public length(): number {
    return this.foldr((acc: number, _: any) => acc + 1, 0)
  }

  public map<T>(transform: (e: T) => T): List {
    return this.foldr((acc: List, curr: T) => List.create(transform(curr), ...acc.values), List.create())
  };

  public foldl<T, U>(f: (acc: U, e: T) => U, acc: U): U {
    this.forEach(e => acc = f(acc, e as T))
    return acc
  };
  
  public foldr<T, U>(f: (acc: U, e: T) => U, acc: U): U {
    this.reverse()
    this.forEach(e => acc = f(acc, e as T))
    return acc
  };

  public reverse(): List {
    for (let index = 0; index < this.values.length / 2; index++) {
      let tmp = this.values[index]
      let indexToSwapWith = this.values.length - 1 - index;
      this.values[index] = this.values[indexToSwapWith]
      this.values[indexToSwapWith] = tmp
    }
    return this
  }
}

File: go/list-ops/list_ops.go
package listops

type IntList = slice[int]
type slice[T any] []T

func (s slice[T]) Foldl(fn func(T, T) T, initial T) T {
	if s.Length() == 0 {
		return initial
	}
	return s[1:].Foldl(fn, fn(initial, s[0]))
}

func (s slice[T]) Foldr(fn func(T, T) T, initial T) T {
	if s.Length() == 0 {
		return initial
	}
	max := s.Length() - 1
	return s[:max].Foldr(fn, fn(s[max], initial))
}

func (s slice[T]) Filter(fn func(T) bool) slice[T] {
	if s.Length() == 0 {
		return s
	}
	res := slice[T]{}
	if fn(s[0]) {
		res = res.Append(slice[T]{s[0]})
	}
	return res.Append(s[1:].Filter(fn))
}

func (s slice[T]) Length() (cnt int) {
	cnt = 0
	for range s {
		cnt += 1
	}
	return
}

func (s slice[T]) Map(fn func(T) T) slice[T] {
	if s.Length() == 0 {
		return s
	}
	return slice[T]{fn(s[0])}.Append(s[1:].Map(fn))
}

func (s slice[T]) Reverse() slice[T] {
	if s.Length() <= 1 {
		return s
	}
	return s[1:].Reverse().Append(slice[T]{s[0]})
}

func (s slice[T]) Append(lst slice[T]) slice[T] {
	if s.Length() == 0 {
		return lst
	}
	res := make(slice[T], s.Length()+lst.Length())
	copy(res[:s.Length()], s)
	copy(res[s.Length():], lst)
	return res
}

func (s slice[T]) Concat(lists []slice[T]) (res slice[T]) {
	res = s
	for _, list := range lists {
		res = res.Append(list)
	}
	return
}
